### PR TITLE
Implement lazy permutations and intersection check

### PR DESCRIPTION
**Pull request by flyingsilverfin**
## What is the goal of this PR?

We impement lazy permutation generation and a short-circuiting intersection existence check.

## What are the changes implemented in this PR?

* create new `collections/Permutations` which contains a lazy permutation generator and a non-lazy permutation generator
  * lazy permutation generator follows C++ `next_permutation` algorithm - explanation at https://stackoverflow.com/questions/352203/generating-permutations-lazily
* implement an optimised intersection existence check which doesn't require computing the entire intersection set

### CODE REVIEW

**Comment by grabl** [comment]
### PR Review Checklist

Do not edit the content of this comment. The PR reviewer should simply update this comment by ticking each review item below, as they get completed.

----
**Code**

- [ ] Packages, classes, and methods have a single domain of responsibility.
- [ ] Packages, classes, and methods are grouped into cohesive and consistent domain model.
- [ ] The code is canonical and the minimum required to achieve the goal.
- [ ] Modules, libraries, and APIs are easy to use, robust (foolproof and not errorprone), and tested.
- [ ] Logic and naming has clear narrative that communicates the accurate intent and responsibility of each module (e.g. method, class, etc.).
- [ ] The code is algorithmically efficient and scalable for the whole application.

**Architecture**

- [ ] Any required refactoring is completed, and the architecture does not introduce technical debt incidentally.
- [ ] Any required build and release automations are updated and/or implemented.
- [ ] Any new components follows a consistent style with respect to the pre-existing codebase.
- [ ] The architecture intuitively reflects the application domain, and is easy to understand.
- [ ] The architecture has a well-defined hierarchy of encapsulated components.
- [ ] The architecture is extensible and scalable.

File: collection/Collections.java [hunk 1]
```diff
@@ -135,12 +111,28 @@ public static <T> Set<T> intersection(Set<T> set1, Set<T> set2) {
...
         }
         Set<T> intersection = new HashSet<>();
-        for (T elem: minSet) {
+        for (T elem : minSet) {
             if (maxSet.contains(elem)) intersection.add(elem);
         }
         return intersection;
     }
 
+    public static <T> boolean hasIntersection(Set<T> set1, Set<T> set2) {
```
Comment (flyingsilverfin) [inline collection/Collections.java:120]: a short-circuiting intersection check for regular sets

File: collection/Permutations.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+import static java.util.Collections.emptyIterator;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singletonList;
+
+public class Permutations {
+
+    /**
+     * We implement the C++ STL next_permutation method of lazily generating permutations
+     */
+    public static <T> Iterator<List<T>> permutations(Set<T> items) {
```
Comment (flyingsilverfin) [inline collection/Permutations.java:23]: return a lazy iterator of permutations

File: collection/Permutations.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        // assign a comparable ordering over the items
+        Map<Integer, T> mapping = new HashMap<>();
+        List<T> sortedItems = items.stream().sorted(Comparator.comparing(Object::toString)).collect(Collectors.toList());
+        int[] itemKeys = new int[sortedItems.size()];
+        for (int i = 0; i < sortedItems.size(); i++) {
+            mapping.put(i, sortedItems.get(i));
+            itemKeys[i] = i;
+        }
+
+        return new Iterator<List<T>>() {
```
Comment (haikalpribadi) [inline collection/Permutations.java:35]: are we sure this should be an anonymous class? seems pretty involved and significant?

**Review by haikalpribadi — APPROVED** [review]

**Pull request merged**